pkg/providers/vault/client.go

```diff
@@ -45,11 +46,41 @@ func NewSecretClient(config SecretConfig) (pkg.SecretClient, error) {
 
 }
 
-// GetValues retrieves the secrets at the provided path that match the specified keys.
+// GetSecrets retrieves the secrets at the provided path that match the specified keys.
 func (c Client) GetSecrets(path string, keys ...string) (map[string]string, error) {
-	data, err := c.getAllKeys(path)
-	if err != nil {
-		return nil, err
+	data := make(map[string]string)
+	var err error
+	addRetryAttempts := c.HttpConfig.AdditionalRetryAttempts
+	switch {
+	case addRetryAttempts < 0:
+		return nil, pkg.NewErrSecretStore(fmt.Sprintf("invalid retry attempts setting %d", addRetryAttempts))
+	case addRetryAttempts == 0:
+		// no retries
+		data, err = c.getAllKeys(path)
+		if err != nil {
+			return nil, err
+		}
+	case addRetryAttempts > 0:
```

Comment:
I think this retry logic should include the logic which is checking for the presence of the specified keys:
https://github.com/edgexfoundry/go-mod-secrets/blob/434d16d649e47d651ba8f79a40d34e3f4ad9d5d7/pkg/providers/vault/client.go#L91-L106

I think this will make out solution to the race condition cover more scenarios, as this will also address the situation where some of the keys are populated, but not all of them. For example, as this is now we will enter the retry loop if we have an HTTP(s) error. From  my little experience with Vault seems to be directly related to auth issues or no secrets being populated at the provided path. However, if the security bootstrapping process creates one of the secrets, but not both, we will throw an error here:
https://github.com/edgexfoundry/go-mod-secrets/blob/434d16d649e47d651ba8f79a40d34e3f4ad9d5d7/pkg/providers/vault/client.go#L105

and the service will exit.

Reply:
I have a different take on this proposed changed.  The Vault API stores maps at each path.  According to the [Vault API](https://www.vaultproject.io/api/secret/kv/kv-v1.html#create-update-secret) you set the key/value pairs all at once.  I don't see the scenario where "password" has been populated but "username" as not, for example, because when you PUT the secret you must specify both at the same time.

Reply:
So these suggested changes are not needed? If that is so then ignore my comment.

Reply:
To clarify, "refactor retry loop to GetSecrets instead of getAllKeys" that seems fine to me to move the retry logic higher in the stack.  To the point of checking individual values in the secret (aka `append(notFound, key)`), my opinion is that it will add negligible additional robustness.